src/exercises/day-3/safe-ffi-wrapper.md

```diff
@@ -10,13 +10,36 @@ You will want to consult the manual pages:
 * [`readdir(3)`](https://man7.org/linux/man-pages/man3/readdir.3.html)
 * [`closedir(3)`](https://man7.org/linux/man-pages/man3/closedir.3.html)
 
-You will also want to browse the [`std::ffi`] module, particular for [`CStr`]
-and [`CString`] types which are used to hold NUL-terminated strings coming from
-C. The [Nomicon] also has a very useful chapter about FFI.
+You will also want to browse the [`std::ffi`] module. There you find a number of
+string types which you need for the exercise:
+
+| Types                      | Encoding       | Use                            |
+|----------------------------|----------------|--------------------------------|
+| [`str`] and [`String`]     | UTF-8          | Text processing in Rust        |
+| [`CStr`] and [`CString`]   | NUL-terminated | Communicating with C functions |
+| [`OsStr`] and [`OsString`] | OS-specific    | Communicating with the OS      |
+
+You will convert between all these types:
+
+- `&str` to `CString`: you need to allocate space for a trailing `\0` character,
+- `CString` to `*const i8`: you need a pointer to call C functions,
+- `*const i8` to `&CStr`: you need something which can find the trailing `\0` character,
+- `&CStr` to `&[u8]`: you will ultimately return a slice of bytes,
+- `&[u8]` to `&OsStr`: you can create an `OsString` from `&[u8]`. Use
+  [`OsStrExt`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html)
+  for this.
+- `&OsStr` to `OsString`: you need to take ownership of the data before you call
```

Comment:
Usually, it's a copy

Reply:
What is a copy? In Rust terms, think it's more precise to say that you get an `OsString` from a `&OsStr` by doing a heap allocation (ownership!) and then copying the data into the new place on the heap.

Reply:
> You don't take ownership of the data. You make an owned copy

Ah... now I get what you mean! You're right that "the data" is a copy after you construct the `OsString`. Let me make a follwup PR.

Reply:
#602 has more changes to this. Please give is a look :smile: